internal/levee.go

```diff
@@ -39,75 +41,210 @@ func run(pass *analysis.Pass) (interface{}, error) {
 	sourcesMap := pass.ResultOf[source.Analyzer].(source.ResultType)
 
-	// Only examine functions that have sources
-	for fn, sources := range sourcesMap {
-		for _, b := range fn.Blocks {
-			if b == fn.Recover {
-				// TODO Handle calls to sinks in a recovery block.
-				continue // skipping Recover since it does not have instructions, rather a single block.
+	for fn := range sourcesMap {
+		analyzeFn(fn, conf, pass)
+	}
+
+	return nil, nil
+}
+
+func analyzeFn(fn *ssa.Function, conf *config.Config, pass *analysis.Pass) {
+	graph := newGraph()
+
+	for _, p := range fn.Params {
+		if conf.IsSource(utils.Dereference(p.Type())) {
+			graph.addSource(p.Name(), p)
+		}
+	}
+
+	// this is needed to capture variables referred to by closures
+	for _, p := range fn.FreeVars {
+		switch t := p.Type().(type) {
+		case *types.Pointer:
+			if s, ok := utils.Dereference(t).(*types.Named); ok && conf.IsSource(s) {
+				graph.addSource(p.Name(), p)
 			}
+		}
+	}
 
-			for _, instr := range b.Instrs {
-				v, ok := instr.(*ssa.Call)
-				if !ok {
-					continue
+	for _, b := range fn.Blocks {
+		for _, inst := range b.Instrs {
+			value, _ := inst.(ssa.Value)
+
+			switch t := inst.(type) {
+			case *ssa.Alloc:
+				if conf.IsSource(utils.Dereference(t.Type())) && !source.IsProducedBySanitizer(t, conf) {
+					graph.addSource(value.Name(), t)
+				}
+
+			case *ssa.Call:
+				parameters := t.Call.Args
+				var parameterNames []string
+				for _, p := range parameters {
+					parameterNames = append(parameterNames, p.Name())
 				}
 				switch {
-				case conf.IsPropagator(v):
-					// Handling the case where sources are propagated to io.Writer
-					// (ex. proto.MarshalText(&buf, c)
-					// In such cases, "buf" becomes a source, and not the return value of the propagator.
-					// TODO Do not hard-code logging sinks usecase
-					// TODO  Handle case of os.Stdout and os.Stderr.
-					// TODO  Do not hard-code the position of the argument, instead declaratively
-					//  specify the position of the propagated source.
-					// TODO  Consider turning propagators that take io.Writer into sinks.
-					if a := getArgumentPropagator(conf, v); a != nil {
-						sources = append(sources, source.New(a, conf))
-					} else {
-						sources = append(sources, source.New(v, conf))
-					}
-
-				case conf.IsSink(v):
-					// TODO Only variadic sink arguments are currently detected.
-					if sinkVarargs := varargs.New(v); sinkVarargs != nil {
-						for _, s := range sources {
-							if sinkVarargs.ReferredBy(s) && !s.IsSanitizedAt(v) {
-								report(pass, s, v)
-								break
-							}
-						}
-					}
+				case conf.IsSink(t):
+					graph.addSink(t, parameterNames)
+				case conf.IsPropagator(t):
+					graph.addSource(value.Name(), t)
+				case conf.IsSanitizer(t):
+					graph.addSanitizer(t)
+				default:
+					graph.addCallEdges(*t, parameterNames)
 				}
+
+			case *ssa.FieldAddr:
+				if conf.IsSourceFieldAddr(t) {
+					graph.addSource(value.Name(), t)
+				}
```

Comment:
This work was already done by the `source` analyzer, which should remain the source of truth for Source identification.